### PR TITLE
Bookmarks undo deletion

### PR DESCRIPTION
**Pull request by mallexxx**
Task/Issue URL: https://app.asana.com/0/1202406491309510/1205306631184883/f

**Description**:
- Undoing bookmarks removal using Cmd+Z

**Steps to test this PR**:
1. Select a bookmark(s) and folder(s) in the Bookmark Manager/Bookmarks Popup/Bookmarks Bar, remove; hit Cmd+z - validate the entities are restored at initial positions
2. Hit Cmd+Shift+Z to redo removal
3. Activate the Address Bar - validate undo is cleared; Activate another tab - validate undo is cleared

**Definition of Done**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)

### CODE REVIEW

File: UnitTests/Bookmarks/Model/LocalBookmarkManagerTests.swift [hunk 1]
```diff
@@ -440,9 +790,141 @@ fileprivate extension Bookmark {
...
+        case let bookmark as Bookmark:
+            Bookmark(id: bookmark.id, url: bookmark.url, title: bookmark.title, isFavorite: bookmark.isFavorite, parentFolderUUID: bookmark.parentFolderUUID, faviconManagement: bookmark.faviconManagement)
+        case let folder as BookmarkFolder:
+            BookmarkFolder(id: folder.id, title: folder.title, parentFolderUUID: folder.parentFolderUUID, children: folder.children)
+        default:
+            fatalError("Unexpected entity type \(self)")
+        }
+    }
+}
+private typealias BookmarksBuilder = ArrayBuilder<BookmarksBuilderItem>
```
Comment (ayoy) [inline UnitTests/Bookmarks/Model/LocalBookmarkManagerTests.swift:864]: I realize that this builder works on `BaseBookmarkEntity` type and I appreciate you adding it, but just wanted to remind and check if you remembered that we have `BookmarkTreeBuilder` in BSK that works on `BookmarkEntity` (managed object class) and could have perhaps be reused here. I understand if it couldn't 👍

**Review by ayoy — CHANGES_REQUESTED** [review]
@mallexxx this is working great, amazing job as always! 🎉 

I've only found that when undoing favorites deletion, we're not putting them back at the original index in favorites array. See the comment inline. It won't be trivial, but also shouldn't be terribly much work. Let me know what you think. cc @jotaemepereira

File: DuckDuckGo/Bookmarks/Model/BookmarkManager.swift [hunk 2]
```diff
@@ -437,6 +483,158 @@ final class LocalBookmarkManager: BookmarkManager {
...
+    deinit {
+        MainActor.assumeIsolated {
+            for id in knownIds {
+                Self.restorableFolders[id] = nil
+            }
+        }
+    }
+}
+enum RestorableBookmarkEntity {
+    case bookmark(url: String, title: String, isFavorite: Bool, parent: KnownBookmarkFolderIdList?, index: Int?)
```
[the remark this answers is not in the document]
Comment (jotaemepereira) [reply DuckDuckGo/Bookmarks/Model/BookmarkManager.swift]: Thanks @ayoy . This was fixed and merged to this branch.

**Review by ayoy — APPROVED** [review]
🚀  thanks a lot @jotaemepereira!

**Pull request merged**